### PR TITLE
gpcheckcat enhancement to report if there are tables created using mix distribution policy 

### PR DESCRIPTION
**Pull request by ravoorsh**
Gpcheckcat is enhanced to report if there are tables created using legacy and non legacy hashops in distribution policy. This check is essential as ORCA fails back to planner if there is mix distribution policy found.

Option Added 
`gpcheckcat -R mix_distribution_policy
`

High level design of "gpcheckcat -R mix_distribution_policy" Check if there are tables created with both legacy and non legacy hashops in distribution policy.

1.      If there is Mix Distribution Policy
-       Report error that there is mix distribution policy.
-       Create an sql file containing the sql query using which the user can
            obtain the table names created with legacy and non legacy hashops

2.     In case there is no Mixed Policy
-      Check if gp_use_legacy_hashops is on and there are no tables created with legacy hashops
 		Report Error
-      Check if gp_use_legacy_hashops is off and there are no tables created with non legacy hashops
 		Report Error

Example run:
```
gpcheckcat -R mix_distribution_policy ravoorsh
Truncated batch size to number of primaries: 4

Connected as user 'ravoorsh' to database 'ravoorsh', port '7000', gpdb version '7.0' ------------------------------------------------------------------- Batch size: 4
Performing test 'mix_distribution_policy'
[ERROR]: Found tables created using both legacy and non legacy hashops in distribution policy. 
Check gpcheckcat.distpolicy.sql file to query the tables. 
Total runtime for test 'mix_distribution_policy': 0:00:00.00

SUMMARY REPORT: FAILED
=================================================================== 
Completed 1 test(s) on database 'ravoorsh' at 2024-03-06 11:22:40 with elapsed time 0:00:00 Failed test(s) that are not reported here: mix_distribution_policy See /Users/ravoorsh/gpAdminLogs/gpcheckcat_20240306.log for detail
```

Test Cases Added:
Both Behave and unit test cases added covering the various scenarios with Mix distribution policy and with no mix distribution policy.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: gpMgmt/bin/gpcheckcat [hunk 1]
```diff
@@ -1997,6 +1997,201 @@ def checkOrphanedToastTables():
...
+        setError(ERROR_NOREPAIR)
+        GV.checkStatus = False
+        myprint('[ERROR] executing test: checkMixDistPolicy')
+        myprint('  Execution error: ' + str(e))
+
+def printDistPolicyMsg(gp_use_legacy_hashops,n_legacy_dist_class, n_total_dist_class):
+
+    GV.checkStatus = True
+
+    if n_legacy_dist_class == 0 and gp_use_legacy_hashops == "on":
```
Comment (kainwen) [inline gpMgmt/bin/gpcheckcat:2178]: The `if` condition here might be better as `if n_total_dist_class - n_legacy_dist_class > 0 and gp_use_legacy_hashops == "on":`. Also the below err LOG might be: 

> GUC gp_use_legacy_hashops is on, all new created table defaultly use legacy hash ops in for hash distributed table, but there are tables using non-legacy hash ops in the cluster.

File: gpMgmt/bin/gpcheckcat [hunk 2]
```diff
@@ -1997,6 +1997,201 @@ def checkOrphanedToastTables():
...
+        myprint(
+            '[ERROR]: GUC gp_use_legacy_hashops is on.'
+            ' All tables created will use legacy hashops in distribution policy'
+            )
+        GV.checkStatus = False
+
+    elif n_legacy_dist_class == 0 and gp_use_legacy_hashops == "off":
+        GV.checkStatus = True
+
+    elif n_legacy_dist_class == n_total_dist_class and gp_use_legacy_hashops == "off" :
```
Comment (kainwen) [inline gpMgmt/bin/gpcheckcat:2188]: Maybe refine the `if condition` to be:

```python
elif n_legacy_dist_class > 0 and gp_use_legacy_hashops == "off" :
```

and refine the err LOG as:

> GUC gp_use_legacy_hashops is off, all new created table defaultly use non-legacy hash ops in for hash distributed table, but there are tables using legacy hash ops in the cluster.

**Review by kainwen — COMMENTED** [review]
Overall LGTM, some minor comments. Thanks for the work!

Please wait for one team mate from CM team to look into this (plus the cases).

**Comment by nimish350** [comment]
Might be something minor but when testing this I noticed that after I run the `gpcheckcat.distpolicy.sql` file that is generated, the output does not have heading on which tables are for which type of policies

For example:
```
navarsekar=# \i gpcheckcat.distpolicy.sql 
  localoid   
-------------
 test_heap21
 test_heap22
 test_heap23
 test_heap24
 test_heap25

  localoid   
-------------
 test_heap1
 test_heap2
 test_heap3
 test_heap4
 test_heap5
```

Instead of the above, it can be something like: 
```
 Legacy Policy   
-------------
 test_heap21
 test_heap22
 test_heap23
 test_heap24
 test_heap25

Non-Legacy Policy   
-------------
 test_heap1
 test_heap2
 test_heap3
 test_heap4
 test_heap5

File: gpMgmt/bin/gpcheckcat [hunk 3]
```diff
@@ -1997,6 +1997,201 @@ def checkOrphanedToastTables():
...
+                        gp_use_legacy_hashops = fetch_guc_value("gp_use_legacy_hashops")
+                        printDistPolicyMsg(gp_use_legacy_hashops,
+                                         n_legacy_dist_class,
+                                         n_total_dist_class
+                                         )
+
+    except Exception as e:
+        setError(ERROR_NOREPAIR)
+        GV.checkStatus = False
+        myprint('[ERROR] executing test: checkMixDistPolicy')
```
Comment (shrakesh) [inline gpMgmt/bin/gpcheckcat:2171]: this should be test name inplace of function name
nit:checkMixDistPolicy -> mix_distribution_policy

File: gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckcat.py [hunk 4]
```diff
@@ -441,6 +441,164 @@ def count_joins():
...
+    @patch('gpcheckcat.connect2')
+    def test_fetch_guc_value_exception_on_cursor_enter(self, mock_connect2):
+        # Mocking the database connection to raise an exception when entering the cursor context
+        mock_connect2.return_value.cursor.return_value.__enter__.side_effect = Exception("Simulated error entering cursor context")
+
+        # Call the function to test
+        self.subject.fetch_guc_value("gp_use_legacy_hashops")
+
+        # Assertions
+        self.assertEqual(mock_connect2.call_count, 1)
```
Comment (shrakesh) [inline gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckcat.py:520]: if we are expecting exception at connect2 function in that case we should assert the call count for curs.execute (should be 0),  curs.fetchone(should be 0), setError(should be 1) and myprint(should be 2) 
just to make sure that the code flow is as expected after the exception. 

we can also check the return value of function fetch_guc_value , it should be None.

File: gpMgmt/bin/gpcheckcat (hunk 1 above)
Comment (ravoorsh) [reply gpMgmt/bin/gpcheckcat]: Done

File: gpMgmt/bin/gpcheckcat (hunk 2 above)
Comment (ravoorsh) [reply gpMgmt/bin/gpcheckcat]: Done.

File: gpMgmt/bin/gpcheckcat (hunk 3 above)
Comment (ravoorsh) [reply gpMgmt/bin/gpcheckcat]: Done

File: gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckcat.py [hunk 5]
```diff
@@ -441,6 +441,164 @@ def count_joins():
...
+   localoid :: regclass :: text
+from
+   gp_distribution_policy,
+   legacy_opclass_oids
+where
+   policytype = 'p'
+   and not (distclass :: oid[] && oid_array);
+           '''
+        self.assertEqual.__self__.maxDiff = None
+        self.assertTrue(generated_query.strip(), expected_query.strip())
```
[the remark this answers is not in the document]
Comment (ravoorsh) [reply gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckcat.py]: Yeah, This was a typo, corrected the same

**Comment by ravoorsh** [comment]
@nimish350 Changed the query to display as follows
  Non Legacy Policy
----------------------
 t_new
 t1_new
 t_random_new
 rank_new

    Legacy Policy
----------------------
 t_old
 t1_old
 rank_old
 rank_old_1_prt_extra

**Review by kainwen — APPROVED** [review]
Code change LGTM.

File: gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckcat.py [hunk 6]
```diff
@@ -441,6 +441,177 @@ def count_joins():
...
+
+        mock_cursor = Mock()
+        mock_connect2.return_value.cursor.return_value.__enter__.return_value = mock_cursor
+        mock_cursor.execute.side_effect = Exception("Simulated error during execution")
+
+        # Call the function to test
+        self.subject.fetch_guc_value("gp_use_legacy_hashops")
+
+        # Assertions
+        self.assertEqual(mock_cursor.execute.call_count, 1)
```
Comment (shrakesh) [inline gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckcat.py:496]: duplicate lines. 496 and 497

**Review by shrakesh — APPROVED** [review]
Approving with minor clean up comment. 
LGTM

**Comment by nimish350** [comment]
Recommended changing the error message to: 
`GUC gp_use_legacy_hashops is off. all newly created tables will use non legacy hash ops by default for hash distributed table, but there are tables using legacy hash ops in the cluster. Please run the "gpcheckcat.distpolicy.sql" file to list the tables. `

**Comment by nimish350** [comment]
LGTM +1

```
navarsekar=# \i gpcheckcat.distpolicy.sql
 Legacy Policy  
----------------
 testand_heap11
 testand_heap12
 testand_heap13
 testand_heap14
 testand_ao11
 testand_ao12
 testand_ao13
 testand_ao14
 testand_aoco11
 testand_aoco12
 testand_aoco13
 testand_aoco14
(12 rows)

 Non Legacy Policy 
-------------------
 testand_heap1
 testand_heap2
 testand_heap3
 testand_heap4
 testand_heap5
 testand_heap6
 testand_heap7
 testand_heap8
 testand_heap9
 testand_heap10
 testand_ao1
 testand_ao2
 testand_ao3
 testand_ao4
 testand_ao5
 testand_ao6
 testand_ao7
 testand_ao8
 testand_ao9
 testand_ao10
 testand_aoco1
 testand_aoco2
 testand_aoco3
 testand_aoco4
 testand_aoco5
 testand_aoco6
 testand_aoco7
 testand_aoco8
 testand_aoco9
 testand_aoco10
(30 rows)
```

gpcheckcat run:
```
Truncated batch size to number of primaries: 4

Connected as user 'navarsekar' to database 'navarsekar', port '7000', gpdb version '7.1'
-------------------------------------------------------------------
Batch size: 4
Performing test 'mix_distribution_policy'
[ERROR]: Found tables created using both legacy and non legacy hashops in distribution policy.Please run the gpcheckcat.distpolicy.sql file to list the tables.
Total runtime for test 'mix_distribution_policy': 0:00:00.02

SUMMARY REPORT: FAILED
===================================================================
```

**Review by nimish350 — APPROVED** [review]

**Comment by ravoorsh** [comment]
> Hi, one more question: does `gpcheckcat` (without cmd args) will try this check defaultly? Can we not trigger it defaultlly? Thanks.

Yes, gpcheckcat without any options will trigger this test

```
#gpcheckcat
Truncated batch size to number of primaries: 4

Connected as user 'ravoorsh' to database 'template1', port '7000', gpdb version '7.1'
-------------------------------------------------------------------
Batch size: 4
Performing test 'unique_index_violation'
Total runtime for test 'unique_index_violation': 0:00:01.34
Performing test 'duplicate'
Total runtime for test 'duplicate': 0:00:05.14
Performing test 'missing_extraneous'
Total runtime for test 'missing_extraneous': 0:00:00.69
Performing test 'inconsistent'
Total runtime for test 'inconsistent': 0:00:04.50
Performing test 'foreign_key'
Total runtime for test 'foreign_key': 0:00:01.65
Performing test 'pgclass'
Total runtime for test 'pgclass': 0:00:00.05
Performing test 'namespace'
Total runtime for test 'namespace': 0:00:00.01
Performing test 'distribution_policy'
Total runtime for test 'distribution_policy': 0:00:00.02
Performing test 'dependency'
Total runtime for test 'dependency': 0:00:00.45
Performing test 'part_integrity'
Total runtime for test 'part_integrity': 0:00:00.05
Performing test 'orphaned_toast_tables'
Total runtime for test 'orphaned_toast_tables': 0:00:00.10
Performing test 'aoseg_table'
Total runtime for test 'aoseg_table': 0:00:00.00
Performing test 'ao_lastrownums'
Total runtime for test 'ao_lastrownums': 0:00:00.00
Performing test 'mix_distribution_policy'
Total runtime for test 'mix_distribution_policy': 0:00:00.00

SUMMARY REPORT: PASSED
===================================================================
Completed 14 test(s) on database 'template1' at 2024-03-19 07:43:29 with elapsed time 0:00:14
Found no catalog issue
```

**Comment by kainwen** [comment]
> > Hi, one more question: does `gpcheckcat` (without cmd args) will try this check defaultly? Can we not trigger it defaultlly? Thanks.
> 
> Yes, gpcheckcat without any options will trigger this test
> 
> ```
> #gpcheckcat
> Truncated batch size to number of primaries: 4
> 
> Connected as user 'ravoorsh' to database 'template1', port '7000', gpdb version '7.1'
> -------------------------------------------------------------------
> Batch size: 4
> Performing test 'unique_index_violation'
> Total runtime for test 'unique_index_violation': 0:00:01.34
> Performing test 'duplicate'
> Total runtime for test 'duplicate': 0:00:05.14
> Performing test 'missing_extraneous'
> Total runtime for test 'missing_extraneous': 0:00:00.69
> Performing test 'inconsistent'
> Total runtime for test 'inconsistent': 0:00:04.50
> Performing test 'foreign_key'
> Total runtime for test 'foreign_key': 0:00:01.65
> Performing test 'pgclass'
> Total runtime for test 'pgclass': 0:00:00.05
> Performing test 'namespace'
> Total runtime for test 'namespace': 0:00:00.01
> Performing test 'distribution_policy'
> Total runtime for test 'distribution_policy': 0:00:00.02
> Performing test 'dependency'
> Total runtime for test 'dependency': 0:00:00.45
> Performing test 'part_integrity'
> Total runtime for test 'part_integrity': 0:00:00.05
> Performing test 'orphaned_toast_tables'
> Total runtime for test 'orphaned_toast_tables': 0:00:00.10
> Performing test 'aoseg_table'
> Total runtime for test 'aoseg_table': 0:00:00.00
> Performing test 'ao_lastrownums'
> Total runtime for test 'ao_lastrownums': 0:00:00.00
> Performing test 'mix_distribution_policy'
> Total runtime for test 'mix_distribution_policy': 0:00:00.00
> 
> SUMMARY REPORT: PASSED
> ===================================================================
> Completed 14 test(s) on database 'template1' at 2024-03-19 07:43:29 with elapsed time 0:00:14
> Found no catalog issue
> ```

Can we change it to not trigger defaultly?
Also want to hear team's idea @nimish350 @shrakesh 
Thanks.

**Comment by nimish350** [comment]
> > > Hi, one more question: does `gpcheckcat` (without cmd args) will try this check defaultly? Can we not trigger it defaultlly? Thanks.
> > 
> > 
> > Yes, gpcheckcat without any options will trigger this test
> > ```
> > #gpcheckcat
> > Truncated batch size to number of primaries: 4
> > 
> > Connected as user 'ravoorsh' to database 'template1', port '7000', gpdb version '7.1'
> > -------------------------------------------------------------------
> > Batch size: 4
> > Performing test 'unique_index_violation'
> > Total runtime for test 'unique_index_violation': 0:00:01.34
> > Performing test 'duplicate'
> > Total runtime for test 'duplicate': 0:00:05.14
> > Performing test 'missing_extraneous'
> > Total runtime for test 'missing_extraneous': 0:00:00.69
> > Performing test 'inconsistent'
> > Total runtime for test 'inconsistent': 0:00:04.50
> > Performing test 'foreign_key'
> > Total runtime for test 'foreign_key': 0:00:01.65
> > Performing test 'pgclass'
> > Total runtime for test 'pgclass': 0:00:00.05
> > Performing test 'namespace'
> > Total runtime for test 'namespace': 0:00:00.01
> > Performing test 'distribution_policy'
> > Total runtime for test 'distribution_policy': 0:00:00.02
> > Performing test 'dependency'
> > Total runtime for test 'dependency': 0:00:00.45
> > Performing test 'part_integrity'
> > Total runtime for test 'part_integrity': 0:00:00.05
> > Performing test 'orphaned_toast_tables'
> > Total runtime for test 'orphaned_toast_tables': 0:00:00.10
> > Performing test 'aoseg_table'
> > Total runtime for test 'aoseg_table': 0:00:00.00
> > Performing test 'ao_lastrownums'
> > Total runtime for test 'ao_lastrownums': 0:00:00.00
> > Performing test 'mix_distribution_policy'
> > Total runtime for test 'mix_distribution_policy': 0:00:00.00
> > 
> > SUMMARY REPORT: PASSED
> > ===================================================================
> > Completed 14 test(s) on database 'template1' at 2024-03-19 07:43:29 with elapsed time 0:00:14
> > Found no catalog issue
> > ```
> 
> Can we change it to not trigger defaultly? Also want to hear team's idea @nimish350 @shrakesh Thanks.

Can we keep this on by default? There should be no harm in running this test during maintenance windows considering it takes a small amount of time.

**Comment by ashwinstar** [comment]
> Can we keep this on by default? There should be no harm in running this test during maintenance windows considering it takes a small amount of time. [Especially after an upgrade to GPDB 6.x.]

In main I am fine having this ON by default. Though for 6X_STABLE having the test run by default will unnecessarily now cause alarm to users if they have this setup whereas was fine with previous gpcheckcat. It's not about time it takes but the noise it causes to existing environments. This is in similar bucket to pg_description change where its causing issues in existing installs. If gpcheckcat had WARNING feature as we had discussed and this would have just WARNED and not actually failed the catalog check script then it was okay to be shipped as default. Thoughts?

**Comment by nimish350** [comment]
> > Can we keep this on by default? There should be no harm in running this test during maintenance windows considering it takes a small amount of time. [Especially after an upgrade to GPDB 6.x.]
> 
> In main I am fine having this ON by default. Though for 6X_STABLE having the test run by default will unnecessarily now cause alarm to users if they have this setup whereas was fine with previous gpcheckcat. It's not about time it takes but the noise it causes to existing environments. This is in similar bucket to pg_description change where its causing issues in existing installs. If gpcheckcat had WARNING feature as we had discussed and this would have just WARNED and not actually failed the catalog check script then it was okay to be shipped as default. Thoughts?


Fine by me. We can hide the test in the -R flag and by default do not run it in 6.x. @ravoorsh

**Comment by ravoorsh** [comment]
Updated 6x PR.

**Pull request merged**